src/main/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImpl.kt

```diff
@@ -0,0 +1,88 @@
+package com.example.iotmanagementdevice.repository
+
+import com.example.iotmanagementdevice.model.MongoDevice
+import com.example.iotmanagementdevice.model.MongoUser
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.core.MongoTemplate
+import org.springframework.data.mongodb.core.aggregate
+import org.springframework.data.mongodb.core.aggregation.Aggregation.match
+import org.springframework.data.mongodb.core.aggregation.Aggregation.newAggregation
+import org.springframework.data.mongodb.core.aggregation.AggregationPipeline
+import org.springframework.data.mongodb.core.aggregation.Fields
+import org.springframework.data.mongodb.core.aggregation.LookupOperation
+import org.springframework.data.mongodb.core.query.Criteria.where
+import org.springframework.data.mongodb.core.query.Query
+import org.springframework.data.mongodb.core.query.Update
+import org.springframework.data.mongodb.core.query.isEqualTo
+import org.springframework.stereotype.Repository
+
+@Repository
+class UserRepositoryImpl(private val mongoTemplate: MongoTemplate) :
+    UserRepository {
+    override fun findById(id: String): MongoUser? {
+        val query = Query(where(Fields.UNDERSCORE_ID).isEqualTo(id))
+        return mongoTemplate.findOne(query, MongoUser::class.java)
+    }
+
+    override fun findAll(): List<MongoUser> =
+        mongoTemplate.findAll(MongoUser::class.java)
+
+    override fun assignDeviceToUser(userId: String, deviceId: String): Boolean {
+        val userUpdateResult = mongoTemplate.updateFirst(
+            Query(where(Fields.UNDERSCORE_ID).isEqualTo(userId)),
+            Update().addToSet(MongoUser::devices.name, ObjectId(deviceId)),
+            MongoUser::class.java
+        )
+
+        val deviceUpdateResult = mongoTemplate.updateFirst(
+            Query(where(Fields.UNDERSCORE_ID).isEqualTo(deviceId)),
+            Update().set(MongoDevice::userId.name, ObjectId(userId)),
+            MongoDevice::class.java
+        )
+
+        return userUpdateResult.modifiedCount > 0 && deviceUpdateResult.modifiedCount > 0
+    }
+
+    override fun findDevicesByUserId(userId: String): List<MongoDevice> {
```

Comment:
since you have userId in device collection, you can do this without aggregation

Reply:
yeah I know, just do it for practice